follower/follower_builder.go

```diff
@@ -0,0 +1,723 @@
+// updated by calling network.SetIDs.
+func (builder *ObserverServiceBuilder) initNetwork(nodeID module.Local,
+	networkMetrics module.NetworkMetrics,
+	middleware network.Middleware,
+	topology network.Topology,
+	receiveCache *netcache.ReceiveCache,
+) (*p2p.Network, error) {
+
+	codec := cborcodec.NewCodec()
+
+	// creates network instance
+	net, err := p2p.NewNetwork(
+		builder.Logger,
+		codec,
+		nodeID,
+		func() (network.Middleware, error) { return builder.Middleware, nil },
+		topology,
+		p2p.NewChannelSubscriptionManager(middleware),
+		networkMetrics,
+		builder.IdentityProvider,
+		receiveCache,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize network: %w", err)
+	}
+
+	return net, nil
+}
+
+func publicNetworkMsgValidators(log zerolog.Logger, idProvider id.IdentityProvider, selfID flow.Identifier) []network.MessageValidator {
+	return []network.MessageValidator{
+		// filter out messages sent by this node itself
+		validator.ValidateNotSender(selfID),
+		validator.NewAnyValidator(
+			// message should be either from a valid staked node
+			validator.NewOriginValidator(
+				id.NewIdentityFilterIdentifierProvider(filter.IsValidCurrentEpochParticipant, idProvider),
+			),
+			// or the message should be specifically targeted for this node
+			validator.ValidateTarget(log, selfID),
+		),
+	}
+}
+
+// BootstrapIdentities converts the bootstrap node addresses and keys to a Flow Identity list where
+// each Flow Identity is initialized with the passed address, the networking key
+// and the Node ID set to ZeroID, role set to Access, 0 stake and no staking key.
+func BootstrapIdentities(addresses []string, keys []string) (flow.IdentityList, error) {
+
+	if len(addresses) != len(keys) {
+		return nil, fmt.Errorf("number of addresses and keys provided for the boostrap nodes don't match")
+	}
+
+	ids := make([]*flow.Identity, len(addresses))
+	for i, address := range addresses {
+		key := keys[i]
+
+		// json unmarshaller needs a quotes before and after the string
+		// the pflags.StringSliceVar does not retain quotes for the command line arg even if escaped with \"
+		// hence this additional check to ensure the key is indeed quoted
+		if !strings.HasPrefix(key, "\"") {
+			key = fmt.Sprintf("\"%s\"", key)
+		}
+		// networking public key
+		var networkKey encodable.NetworkPubKey
+		err := json.Unmarshal([]byte(key), &networkKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// create the identity of the peer by setting only the relevant fields
+		ids[i] = &flow.Identity{
+			NodeID:        flow.ZeroID, // the NodeID is the hash of the staking key and for the public network it does not apply
+			Address:       address,
+			Role:          flow.RoleAccess, // the upstream node has to be an access node
+			NetworkPubKey: networkKey,
+		}
+	}
+	return ids, nil
+}
+
+func WithSkipValidations(builder *ObserverServiceBuilder) *ObserverServiceBuilder {
```

Comment:
The idiomatic way to do this in flow-go is to use an Options/Config approach similar to `WithBootStrapPeers`

Reply:
Fix: https://github.com/GetElastech/flow-go/commit/db8e7f7e4ffcf04c8234dab7a989026a363083b0